CONSOLE-4265: Add new API field to ConsolePlugin CRD for allowing additional CSP sources

jhadvig opened the pull request:
When serving Console HTML index page, we generate the policy that includes allowed (trustworthy) [sources](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources).
It may be necessary for some dynamic plugins to add new sources in order to avoid CSP violations at Console runtime.
We need to extend the ConsolePlugin CRD, in order to give the plugin creators a mechanism how to specify these source.

We discussed with @vojtechszocs two possible implementation:
1. Add a new field to spec for specifying an array of sources which would apply for all the CSP directives - `spec.csp.allowedSources`. This implementation is pretty straight-forward but automatically applying the same source across different directives could unintentionally block legitimate content or reduce security.

    Example:
```yaml
kind: ConsolePlugin
metadata:
  name: my-console-plugin
spec:
  displayName: "My Custom Console Plugin"
  backend:
    service:
      name: "plugin-backend-service"
      namespace: "plugin-namespace"
      port: 8080
  csp:
    allowedSources:
      - "https://trusted-images.com"
      - "https://cdn.images.com"
```

2. (Proposed solution) Add a new field to spec for specifying an array of sources for each of the directives. Letting users manage different whitelists for different types of content, gives them more flexibility in controlling security settings.

    Example:
```yaml
kind: ConsolePlugin
metadata:
  name: my-console-plugin
spec:
  displayName: "My Custom Console Plugin"
  backend:
    service:
      name: "plugin-backend-service"
      namespace: "plugin-namespace"
      port: 8080
  csp:
    - directive: script-src
      sources:
        - "https://trusted-scripts.com"
    - directive: img-src
      sources:
        - "https://trusted-images.com"
        - "https://cdn.images.com"
    - directive: style-src
      sources:
        - "https://trusted-styles.com"
```

Note: The PR is missing some descriptions and validation and unit tests. Will address those once we align on the API structure.

/assign @spadgett @vojtechszocs 

Story: https://issues.redhat.com/browse/CONSOLE-4265

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,26 @@ type ConsolePluginSpec struct {
+	ImgSrc     DirectiveType = "img-src"
+	StyleSrc   DirectiveType = "style-src"
+	FontSrc    DirectiveType = "font-src"
+	BaseUri    DirectiveType = "base-uri"
```
We don't want plugins to be able to change this.

jhadvig replied:
@spadgett not sure how they could, since these are enums?

spadgett replied:
Right, we should remove `base-uri` from the enums. It's not something we want to expose to plugins.

jhadvig replied:
Oh sorry, I've misunderstood the comment. 
Yes I agree that `base-uri` directive should be only set by console backend.

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,26 @@ type ConsolePluginSpec struct {
+type DirectiveType string
+
+const (
+	DefaultSrc DirectiveType = "default-src"
+	ScriptSrc  DirectiveType = "script-src"
+	ImgSrc     DirectiveType = "img-src"
+	StyleSrc   DirectiveType = "style-src"
+	FontSrc    DirectiveType = "font-src"
```
We should describe what these are or link to documentation. A link to the MDN topic would be good (unless we have a policy against linking to 3rd party docs).

https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP

jhadvig replied:
From what I see we are using links for 3rd party docs in the API - https://github.com/openshift/api/blob/master/config/v1/types_dns.go#L79-L81

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,26 @@ type ConsolePluginSpec struct {
+// ConsolePluginCSP
+type ConsolePluginCSP struct {
+	Directive DirectiveType `json:"directive"`
+	Sources   []string      `json:"source"`
```
I would make this something generic like `values` since we might have directives in the future that aren't sources.

spadgett replied:
Need to make clear that this adds to the existing console directive values rather than replacing them.

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,26 @@ type ConsolePluginSpec struct {
 	// +optional
 	I18n ConsolePluginI18n `json:"i18n"`
+	// csp is a list of Content Security Policy directives for the plugin.
+	CSP []ConsolePluginCSP `json:"csp"`
```
I would lean toward spelling it out: `contentSecurityPolicy`

jhadvig replied:
Definitely 👍

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,26 @@ type ConsolePluginSpec struct {
 	// i18n is the configuration of plugin's localization resources.
 	// +optional
 	I18n ConsolePluginI18n `json:"i18n"`
+	// csp is a list of Content Security Policy directives for the plugin.
```
For the final doc, we'll want more detail about what this is and why you'd need to change it. Examples would be helpful.

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,26 @@ type ConsolePluginSpec struct {
+
+// ConsolePluginCSP
+type ConsolePluginCSP struct {
+	Directive DirectiveType `json:"directive"`
```
I like the idea of making this generic to support other directives as long as we have validation that only the ones we want can be customized. (Plugins shouldn't be able to set any directive.)

jhadvig replied:
@spadgett not sure what you mean here. Could you please be more specific?

spadgett replied:
Just a general comment that I agree with the approach :) But we should be careful what directives we allow plugins to change.

jhadvig replied:
@spadgett this enumeration makes sure that only these types are valid.

spadgett replied:
Right. My point is that we should be deliberate in what we add to this enum in the future.

jhadvig commented:
@spadgett thank you for the review. I've address the comments. PTAL

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,84 @@ type ConsolePluginSpec struct {
+	//     font-src: https://font2.com/
+	//     img-src: https://img1.com/
+	//
+	//   OpenShift web console server CSP response header:
+	//     script-src: https://script1.com/ https://script2.com/ https://script3.com/
+	//     font-src: https://font1.com/ https://font2.com/
+	//     img-src: https://img1.com/
```
These would still include `self` as well which the console sets.

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,84 @@ type ConsolePluginSpec struct {
 	// i18n is the configuration of plugin's localization resources.
 	// +optional
 	I18n ConsolePluginI18n `json:"i18n"`
+	// contentSecurityPolicy is a list of Content Security Policy directives for the plugin.
```
I'd make a comment here that plugins shouldn't normally need to change this unless they are loading assets from somewhere else. I'd also make a comment about how a plugin can test to see if they have any violations.

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,84 @@ type ConsolePluginSpec struct {
+	// Each directive specifies a list of values that indicate server origins and script endpoints
+	// from which the plugin's assets can be loaded.
+	// This helps guard against cross-site scripting (XSS) attacks.
+	// Available directive types are default-src, script-src, img-src, style-src and font-src.
```
```suggestion
	// Available directives are default-src, script-src, img-src, style-src and font-src.
```

spadgett replied:
I'd also note that by default console adds the value `self` for the various `src` directives.

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,84 @@ type ConsolePluginSpec struct {
+	// For more information about the CSP directives, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
+	//
+	// The OpenShift web console server aggregates the CSP directives and values across
+	// all enabled ConsolePlugin CRs, merging them to set a unified CSP header.
```
```suggestion
	// The OpenShift web console server aggregates the CSP directives and values across
	// its own default values and all enabled ConsolePlugin CRs, merging them to set a unified
	// CSP header.
```

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,84 @@ type ConsolePluginSpec struct {
+	// +kubebuilder:validation:Enum:="default-src";"script-src";"img-src";"style-src";"font-src"
+	// +kubebuilder:validation:Required
+	Directive DirectiveType `json:"directive"`
+	// values defines an array of source values mostly specifying server origins
```
```suggestion
	// values defines an array of additional values to append to the console defaults for this directive.
```

We might add directives that take something other than a source in the future.

spadgett commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,84 @@ type ConsolePluginSpec struct {
+
+// ConsolePluginCSP holds configuration for a specific CSP directive
+type ConsolePluginCSP struct {
+	// directive is a type of CSP directive.
```
```suggestion
	// directive specifies which Content-Security-Policy directive to configure.
```

jhadvig commented:
@spadgett thank you for the review. Addressed your comments in the additional commit.
PTAL

jhadvig commented:
/assign @JoelSpeed

spadgett commented:
Thanks, @jhadvig. I agree with the approach and appreciate the detailed doc. Let's have @JoelSpeed take a look 👍

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	// the cluster or if violation reports are observed.
+	// CSP violation reports can be viewed in browser's console during development and testing
+	//  of the plugin in the OpenShift web console.
+	// Available directives are default-src, script-src, img-src, style-src and font-src.
```
Enum values should be PascalCase, `src` here is shorthand for source right? Does it need to be included to make sense?
What if this were 
```suggestion
	// Available directives are Default, Script, Img, Style and Font.
```

jhadvig replied:
+1 for keeping  "source" or "src" in the name, since [CSP are not only sources](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) and making them PascalCase 👍

JoelSpeed replied:
Ack, then `DefaultSrc`, `ScriptSrc` etc seems appropriate. Looking at [CSP docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#directives) it seems there's plenty that wouldn't have `Src` suffixes, we just happen to only be exposing those right now that have the suffix

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	// CSP violation reports can be viewed in browser's console during development and testing
+	//  of the plugin in the OpenShift web console.
+	// Available directives are default-src, script-src, img-src, style-src and font-src.
+	// Each of the available CSP directive may be defined only once in the list.
```
```suggestion
	// Each of the available directives may be defined only once in the list.
```

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	//  of the plugin in the OpenShift web console.
+	// Available directives are default-src, script-src, img-src, style-src and font-src.
+	// Each of the available CSP directive may be defined only once in the list.
+	// By default the console server adds the value 'self'to all the various 'src' directives.
```
```suggestion
	// By default the console server adds the value 'self' to all directives.
```
Reading this I'd wonder by `self` is added? And you say by default, does this mean I can disable this behaviour?

spadgett replied:
Console needs `self` to function for all of these values. It can't be removed.

All of the plugin customizations append to the default console values that console needs to work.

spadgett replied:
Note that in the future we could add directives that don't specify sources (although right now I think everything we expose is a "src" directive). We'd have to come back and update the text here later if we leave that out.

JoelSpeed replied:
So that is to say, if this were a directive that were not a `src` style, then it wouldn't also need the `self` prepending?

spadgett replied:
Right, `self` only makes sense as a value for the various CSP `src` directives.

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	//
+	// Example:
+	//   ConsolePlugin A directives:
+	//     script-src: https://script1.com/ https://script2.com/
```
A string with multiple entries, space separated? Why not a list of strings? I think this is not rendered correctly and we should resolve.

Or are you representing the headers?

jhadvig replied:
this is just an example how to source value are being unified  by the console server. Will add a the final header example as well 👍

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	ContentSecurityPolicy []ConsolePluginCSP `json:"contentSecurityPolicy,omitempty"`
```
By requiring 1 entry, and including the omitempty, you are not making this field discoverable. If the field were rendered in the serialised version with an empty list, would this cause issues for you?

jhadvig replied:
should not cause any issues, but will remove the `omitempty`, since we wanna make the field discoverable.

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	ContentSecurityPolicy []ConsolePluginCSP `json:"contentSecurityPolicy,omitempty"`
```
Are there any fields in the API that exist today that might causes bad interactions when used in conjunction with this new API?

jhadvig replied:
No there are no such a fields.

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+
+// DirectiveType is an enumeration of OpenShift web console supported CSP directives.
+// LoadType is an enumeration of i18n loading types.
+// +kubebuilder:validation:Enum:=default-src;script-src;img-src;style-src;font-src
```
Needs to be PascalCase

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+type ConsolePluginCSP struct {
+	// directive specifies which Content-Security-Policy directive to configure.
+	// Available directive types are default-src, script-src, img-src, style-src and font-src.
+	// +kubebuilder:validation:Enum:="default-src";"script-src";"img-src";"style-src";"font-src"
```
You already have this on the `DirectiveType`, you only need it once

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+// ConsolePluginCSP holds configuration for a specific CSP directive
+type ConsolePluginCSP struct {
+	// directive specifies which Content-Security-Policy directive to configure.
+	// Available directive types are default-src, script-src, img-src, style-src and font-src.
```
Can you expand the godoc here briefly to explain what the directives do, in case the user hasn't explained the higher level field

Eg what is a default vs a script src?

jhadvig replied:
Those are explained as part of the enum definition, together with the link for official docs. Thought that it might be more suitable to explain them briefly there together with link.

JoelSpeed replied:
The enum definition won't be generated into docs, please copy to here and make sure that it ends up in the CRD schema for `oc explain` use later

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	// +kubebuilder:validation:Enum:="default-src";"script-src";"img-src";"style-src";"font-src"
+	// +kubebuilder:validation:Required
+	Directive DirectiveType `json:"directive"`
+	// values defines an array of additional values to append to the console
+	// defaults for this directive.
```
Don't really need additional do we?
```suggestion
	// values defines an array of values to append to the console
	// defaults for this directive.
```

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	// defaults for this directive.
+	// Each ConsolePlugin may define their own directives with their values.
+	// These will be set by the OpenShift web console's backend, as part of
+	// its CSP header.
```
Spell CSP out here I think, since you haven't explained it otherwise in this godoc

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	// its CSP header.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	Values []string `json:"values"`
```
What is the format of this string? Is it always a URL? What is the maximum possible length? What characters are valid, which are not?

spadgett replied:
@JoelSpeed I think we have to leave it open since CSP directive values are not always URLs and don't always follow a specific format.

jhadvig replied:
I agree to leave it open and add additional validation on the backend

spadgett replied:
We probably should disallow `;` since that is a separator in the header between directives. Maybe we can do more validation on the values. Here is the syntax from the standard:

https://www.w3.org/TR/CSP2/#policy-syntax

spadgett replied:
> directive-value = *( WSP / <VCHAR except ";" and ","> )

Looks like `;` and `,` are not allowed in values and whitespace separates values, so can be disallowed as well. `VCHAR` refers to printing characters, so I think pretty much anything else is allowed?

jhadvig replied:
@spadgett this would require following pattern
```go
// +kubebuilder:validation:Pattern=`^[\t\x20!#\$%&'\(\)\*\+\-\.\/0-9:<=>\?@A-Za-z\[\\\]\^_`\{\|\}~]*$`
```

JoelSpeed replied:
Does the regex here not support negative matching? ie, match not this character?

If it doesn't, you could use CEL validations quite easily, `!self.Contains(';')`, which would be readable.

> I don't see a length limit on specific values, although some HTTP servers like Apache have limits on the total size of response headers. If we pick it a limit, it would be an arbitrary value.

Even an arbitrary limit is better than none, how about starting at `1024` and seeing if you get any customers complain that it is too small?

spadgett replied:
> Even an arbitrary limit is better than none, how about starting at 1024 and seeing if you get any customers complain that it is too small?

@JoelSpeed I agree. Works for me.

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=5
+	// +listType=map
+	// +listMapKey=name
```
`name` isn't a field within the ConsolePluginCSP, I think you mean `directive`

jhadvig replied:
Yes 👍

jhadvig commented:
@JoelSpeed thank you for the comments. Addressed them in a separate commit and replied to some of them.

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,91 @@ type ConsolePluginSpec struct {
+	// its Content Security Policy header.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	Values []string `json:"values"`
```
What would be a reasonable/arbitrary length limit for this list, would you expect someone to have more than say, 32 values?

spadgett replied:
32 should be more than enough.

jhadvig replied:
right, I dont think any plugin would ever contribute more then 16

jhadvig commented:
@JoelSpeed comments addressed. PTAL

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	// Each directive specifies a list of values that indicate server origins and script endpoints
+	// from which the plugin's assets can be loaded. This helps guard against cross-site
+	// scripting (XSS) attacks by enforcing strict security policies for asset loading.
+	// Dynamic plugins should to specify this field if they are loading assets form outside
```
```suggestion
	// Dynamic plugins should specify this field if they are loading assets from outside
```

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	// the cluster or if violation reports are observed.
+	// CSP violation reports can be viewed in browser's console during development and testing
+	// of the plugin in the OpenShift web console.
+	// Available directives are default-src, script-src, img-src, style-src and font-src.
```
Needs to be updated to reflect the new names

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	// of the plugin in the OpenShift web console.
+	// Available directives are default-src, script-src, img-src, style-src and font-src.
+	// Each of the available directives may be defined only once in the list.
+	// By default the console server adds the value 'self 'to all the various '*-src' directives.
```
I think we can be more direct in how we state this.
```suggestion
	// The value 'self' will be prepended to all source type directives.
```

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	// For more information about the CSP directives, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
+	//
+	// The OpenShift web console server aggregates the CSP directives and values across
+	// its own default values and all enabled ConsolePlugin CRs, merging them to set a unified
+	// CSP header.
```
Is there a combined maximum that we need to be aware of?

jhadvig replied:
It's recommended to keep CSP headers under 8 KB to ensure compatibility across all browsers, servers, and proxies, since different browsers impose their own limits on HTTP header size. For example:
- Chrome and Edge generally support headers up to 8 KB.
- Firefox and Safari can support headers up to 16 KB or more.

JoelSpeed replied:
In that case, should we impose a total limit in size? You could do this with CEL and a map function that maps every entry in the list to its size and then sum them. We are limiting to 32 items at 1kb each so in theory we could reach 32kb for each of the 5 source types.

I think something like the following on CSP field would work.

```
// +kubebuilder:validation:XValidation:rule="self.map(x, x.values.map(y, y.size()).sum()).sum() < 8192"
```

JoelSpeed replied:
We have the option to restrict the limits down further if you think they can be restricted down, you're the SME on what the quantity of expected values might be here.

And yes, you're right, multiple plugins may contribute, and we may not catch everything with this validation, but, this would at least ensure that the single console plugin doesn't, by itself, violate the restriction. It is the strictest we can be

jhadvig replied:
> And yes, you're right, multiple plugins may contribute, and we may not catch everything with this validation, but, this would at least ensure that the single console plugin doesn't, by itself, violate the restriction. It is the strictest we can be

👍

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	//   OpenShift web console server CSP response header:
+	//     Content-Security-Policy: default-src 'self'; base-uri 'self'; script-src 'self' https://script1.com/ https://script2.com/ https://script3.com/; font-src 'self' https://font1.com/ https://font2.com/; img-src 'self' https://img1.com/; style-src 'self'; frame-src 'none'; object-src 'none'
+	//
+	// +kubebuilder:validation:MinItems=1
```
Drop this one please, it won't work with the discoverability element

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+// ConsolePluginCSP holds configuration for a specific CSP directive
+type ConsolePluginCSP struct {
+	// directive specifies which Content-Security-Policy directive to configure.
+	// Available directive types are DefaultSrc, ScriptSrc, ImgSrc, StyleSrc and FontSrc.
```
Can you include in this godoc the snipper about how to find out more about each of the directives?

jhadvig replied:
will add links for docs for each one

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=32
+	// +kubebuilder:validation:MaxLength=1024
```
I believe the correct syntax is 
```suggestion
	// +kubebuilder:validation:items:MaxLength=1024
```

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=32
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:UniqueItems=true
```
This won't work, you need to use CEL
```suggestion
	// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x == y))",message="each CSP directive value must be unique"
```

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	// +kubebuilder:validation:MaxItems=32
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:UniqueItems=true
+	// +kubebuilder:validation:XValidation:rule=`!self.contains(',')`,message="CSP directive value cannot contain a comma"
+	// +kubebuilder:validation:XValidation:rule=`!self.contains(';')`,message="CSP directive value cannot contain a semicolon"
```
Don't mix quote types please
```suggestion
	// +kubebuilder:validation:XValidation:rule="!self.contains(',')",message="CSP directive value cannot contain a comma"
	// +kubebuilder:validation:XValidation:rule="!self.contains(';')",message="CSP directive value cannot contain a semicolon"
```

JoelSpeed replied:
What about whitespace? You're not excluding whitespace yet?

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	// values defines an array of values to append to the console defaults for this directive.
+	// Each ConsolePlugin may define their own directives with their values.
+	// These will be set by the OpenShift web console's backend, as part of
+	// its Content Security Policy header.
```
You need to explain in the godoc that each item must have at most 1024 chars, cannot contain whitespace, commas or semi-colons.

You also need to explain that at most 32 items can be included in the list.

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,96 @@ type ConsolePluginSpec struct {
+	// +kubebuilder:validation:MaxItems=32
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:UniqueItems=true
+	// +kubebuilder:validation:XValidation:rule=`!self.contains(',')`,message="CSP directive value cannot contain a comma"
+	// +kubebuilder:validation:XValidation:rule=`!self.contains(';')`,message="CSP directive value cannot contain a semicolon"
```
These won't actually work, you'd need a `self.all(x, !x.contains()` style when using CEL on a list like this

jhadvig replied:
will use string type alias, where I will place the validation

JoelSpeed reviewed (commented):
Please make sure when making updates to generate the CRD schema, at the moment I don't think the generation would succeed but you haven't generated so we can't actually see whether it would or not

I'd also like you to write some integration tests (you can see the style in the tests folder in each API group) to check the validations that you are adding actually work and return sensible looking errors

jhadvig commented:
@JoelSpeed addressing most of the comments. The only missing piece is the [size of the directives](https://github.com/openshift/api/pull/2042#discussion_r1791794167), size this will only handle a since ConsolePlugin CR and there could be multiple running on the server.

Adding generated CRD schema + integration test.

PTAL

JoelSpeed commented on console/v1/tests/consoleplugins.console.openshift.io/AAA_ungated.yaml:
```diff
@@ -87,3 +87,75 @@ tests:
+          - directive: ScriptSrc
+            values:
+            - https://script1.com/
+      expectedError: "Duplicate value: map[string]interface {}{\"directive\":\"ScriptSrc\""
```
Do you have the complete error here? Looks like it's been truncated? 

This will do a substring match but I'm curious what comes after this

jhadvig replied:
@JoelSpeed the whole msg is:
`"ConsolePlugin.console.openshift.io \"test-vgzng\" is invalid: spec.contentSecurityPolicy[1]: Duplicate value: map[string]interface {}{\"directive\":\"ScriptSrc\"}"`

JoelSpeed replied:
Ok, we should update this line then to `spec.contentSecurityPolicy[1]: Duplicate value: map[string]interface {}{\"directive\":\"ScriptSrc\"}` to capture the field name and the end that got truncated

JoelSpeed replied:
Outstanding to fix
```suggestion
      expectedError: "spec.contentSecurityPolicy[1]: Duplicate value: map[string]interface {}{\"directive\":\"ScriptSrc\"}"
```

JoelSpeed commented on console/v1/tests/consoleplugins.console.openshift.io/AAA_ungated.yaml:
```diff
@@ -87,3 +87,75 @@ tests:
+          - directive: TestSrc
+            values:
+            - https://script1.com/
+      expectedError: "spec.contentSecurityPolicy[0].directive: Unsupported value: \"TestSrc\": supported values: \"DefaultSrc\", \"ScriptSrc\", \"ImgSrc\", \"StyleSrc\", \"FontSrc\""
```
Please finish all files with an empty new line char

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,118 @@ type ConsolePluginSpec struct {
+// whitespace, commas, or semicolons.
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=1024
+// +kubebuilder:validation:XValidation:rule="!self.contains(' ')",message="CSP directive value cannot contain a whitespace"
```
This covers spaces, do we also need to cover tabs somehow? Does the regex used in CEL support a whitespace grouping? Perhaps using another function like the findAll function would help to find any whitespace using regex and then catch all whitespace in a single rule?

jhadvig replied:
added

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,118 @@ type ConsolePluginSpec struct {
+)
+
+// CSPDirectiveValue is single value for a Content-Security-Policy directive.
+// Each directive value must have a maximum length of 1024 characters and must not contain
+// whitespace, commas, or semicolons.
```
This must go into the godoc for the field that is using this type as well please

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,118 @@ type ConsolePluginSpec struct {
+	// +listType=map
+	// +listMapKey=directive
+	// +optional
+	ContentSecurityPolicy []ConsolePluginCSP `json:"contentSecurityPolicy"`
```
Reminder to add the validation suggested in https://github.com/openshift/api/pull/2042#discussion_r1791708059

JoelSpeed commented on console/v1/tests/consoleplugins.console.openshift.io/AAA_ungated.yaml:
```diff
@@ -87,3 +87,137 @@ tests:
+          - directive: ScriptSrc
+            values:
+            - "*"
+      expectedError: "Invalid value: \"string\": CSP directive value cannot be a wildcard"
```
Please include the field path part of the error

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,122 @@ type ConsolePluginSpec struct {
+	//     Content-Security-Policy: default-src 'self'; base-uri 'self'; script-src 'self' https://script1.com/ https://script2.com/ https://script3.com/; font-src 'self' https://font1.com/ https://font2.com/; img-src 'self' https://img1.com/; style-src 'self'; frame-src 'none'; object-src 'none'
+	//
+	// +kubebuilder:validation:MaxItems=5
+	// +kubebuilder:validation:XValidation:rule="self.map(x, x.values.map(y, y.size()).sum()).sum() < 8192"
```
This needs a message to explain it
```suggestion
	// +kubebuilder:validation:XValidation:rule="self.map(x, x.values.map(y, y.size()).sum()).sum() < 8192",message="the total combined size of values of all directives must not exceed 8192 (8kb)"
```

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,122 @@ type ConsolePluginSpec struct {
+	// values defines an array of values to append to the console defaults for this directive.
+	// Each ConsolePlugin may define their own directives with their values. These will be set
+	// by the OpenShift web console's backend, as part of its Content-Security-Policy header.
+	// The array can contain at most 32 values. Each directive value must have a maximum length
```
```suggestion
	// The array can contain at most 8 values. Each directive value must have a maximum length
```

JoelSpeed replied:
Also need to mention that each value in the array must be unique

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,122 @@ type ConsolePluginSpec struct {
+	// Each ConsolePlugin may define their own directives with their values. These will be set
+	// by the OpenShift web console's backend, as part of its Content-Security-Policy header.
+	// The array can contain at most 32 values. Each directive value must have a maximum length
+	// of 1024 characters and must not contain whitespace, commas, or semicolons.
```
```suggestion
	// of 1024 characters and must not contain whitespace, commas, or semicolons.
	// The value '*' is not permitted.
```

vojtechszocs commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,123 @@ type ConsolePluginSpec struct {
+// +kubebuilder:validation:XValidation:rule="!self.contains(',')",message="CSP directive value cannot contain a comma"
+// +kubebuilder:validation:XValidation:rule="!self.contains(';')",message="CSP directive value cannot contain a semi-colon"
+// +kubebuilder:validation:XValidation:rule="self != '*'",message="CSP directive value cannot be a wildcard"
+type CSPDirectiveValue string
```
Important :exclamation: I think we should explicitly disallow any CSP directive values which are enclosed in single quotes.

`'self'`, `'unsafe-eval'` etc. are all tokens with special meaning when interpreted by the browser and I think that only Console web application itself should be allowed to use these when generating the security policy.

For example, it makes no sense for a plugin to use `'self'` value since that's already included by default. Also, it makes no sense for a plugin to e.g. use `'unsafe-hashes'` or `'unsafe-inline'` to lower the security constraint for specific asset fetch directives.

The above validation list should include a rule that disallows a string that begins and ends with `'` character.

jhadvig replied:
added following validation
```go
// +kubebuilder:validation:XValidation:rule="!(self.startsWith(\"'\") && self.endsWith(\"'\"))",message="CSP directive value cannot start or and with a quote"
```

vojtechszocs commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,123 @@ type ConsolePluginSpec struct {
+	// Each ConsolePlugin may define their own directives with their values. These will be set
+	// by the OpenShift web console's backend, as part of its Content-Security-Policy header.
+	// The array can contain at most 8 values. Each directive value must have a maximum length
+	// of 1024 characters and must not contain whitespace, commas, semicolons. The value '*'is
```
```suggestion
	// of 1024 characters and must not contain whitespace, commas, semicolons. The value '*' or strings enclosed in single quotes are not permitted.
```

yapei commented on console/v1/tests/consoleplugins.console.openshift.io/AAA_ungated.yaml:
```diff
@@ -87,3 +87,137 @@ tests:
+            values:
+            - https://scr ipt1.com/
+      expectedError: "spec.contentSecurityPolicy[0].values[0]: Invalid value: \"string\": CSP directive value cannot contain a whitespace"
+    - name: Should throw an error for invalid CSP directive values with whitespace, using tab
```
given following validation checks, it looks like we also need add more tests here
- name: Should throw an error for invalid CSP directive values with comma
- name: Should throw an error for invalid CSP directive values with semicolon
- name: Should throw an error for invalid CSP directive values with strings enclosed in single quotes(based on vojtechszocs's comment)

jhadvig commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,123 @@ type ConsolePluginSpec struct {
+	// Each directive specifies a list of values that indicate server origins and script endpoints
+	// from which the plugin's assets can be loaded. This helps guard against cross-site
+	// scripting (XSS) attacks by enforcing strict security policies for asset loading.
+	// Dynamic plugins should specify this field if they are loading assets from outside
```
Updated the wording a bit:
```go
	// Dynamic plugins should specify this field if need to load assets from outside
	// the cluster or if violation reports are observed. Dynamic plugins should always prefer
	// loading their assets from within the cluster, either by vendoring them, or fetching
	// from a cluster service.
```

jhadvig commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,123 @@ type ConsolePluginSpec struct {
+	// of the plugin in the OpenShift web console.
+	// Available directive types are DefaultSrc, ScriptSrc, ImgSrc, StyleSrc and FontSrc.
+	// Each of the available directives may be defined only once in the list.
+	// The value 'self' will be prepended to all source type directives.
```
```go
	// The value 'self' is automatically included in all fetch directives by the OpenShift web
	// console's backend.
```
WDYT?

JoelSpeed commented on features/features.go:
```diff
@@ -36,6 +36,13 @@ func AllFeatureSets() map[ClusterProfileName]map[configv1.FeatureSet]*FeatureGat
+					reportProblemsToJiraComponent("Management Console").
+					contactPerson("jhadvig").
+					productScope(ocpSpecific).
+					enableIn(configv1.Default, configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
```
Why add a gate if you are adding it to all sets at once? I think you want to not have it as Default for now

jhadvig replied:
@JoelSpeed hmm but will the filed in this case be available by default? Cause thats what we need.

JoelSpeed replied:
To complete the loop from Slack, it should be introduced as TP and then promoted to Default only once QE are happy that the feature is complete and stable

JoelSpeed commented on features/features.go:
```diff
@@ -36,6 +36,13 @@ func AllFeatureSets() map[ClusterProfileName]map[configv1.FeatureSet]*FeatureGat
+					contactPerson("jhadvig").
+					productScope(ocpSpecific).
+					enableIn(configv1.Default, configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
+					mustRegister()
```
You'll need to add an enhancementPR link for your feature
```suggestion
					enhancementPR("").
					mustRegister()
```

JoelSpeed commented on console/v1/tests/consoleplugins.console.openshift.io/AAA_ungated.yaml:
```diff
@@ -87,3 +87,176 @@ tests:
+          - directive: ScriptSrc
+            values:
+            - https://script1.com/
+      expectedError: "spec.contentSecurityPolicy[1]: Duplicate value: map[string]interface {}{\"directive\":\"ScriptSrc\""
```
I asked you to include the rest of this error message for completeness, but it seems to have gone
It's odd seeing the unbalanced braces

JoelSpeed commented on console/v1/tests/consoleplugins.console.openshift.io/AAA_ungated.yaml:
```diff
@@ -87,3 +87,176 @@ tests:
+          - directive: ScriptSrc
+            values:
+            - "*.example.com"
+    - name: Should throw an error for invalid CSP directive values, starting and ending with quotes
+      initial: |
+        apiVersion: console.openshift.io/v1
+        kind: ConsolePlugin
+        spec:
+          displayName: foo
+          backend:
+            type: Service
+          contentSecurityPolicy:
+          - directive: ScriptSrc
+            values:
+            - "'none'"
+      expectedError: "spec.contentSecurityPolicy[0].values[0]: Invalid value: \"string\": CSP directive value cannot start or and with a quote"
```
What about unbalanced quotes? What if I had
```
- a'bad
- thing'b
```

concatenated this becomes `a'bad thing'b`, is the `'bad thing'` in there problematic?

jhadvig replied:
If a CSP header like `Content-Security-Policy: script-src a'bad thing'b;` is used, the browser will consider it invalid due to the unescaped or unrecognized syntax in a'bad thing'b. The result is that the entire CSP directive could be ignored. That said I think we should disable the quote completely. If you agree will also keep the previous check for a value which starts and ends with quote?

spadgett replied:
Do we allow hash values?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources#hash-algorithm-base64-value

We want to support that for plugins that have inline scripts. It requires using `'` characters, though. (nonce we shouldn't support, however, since there's no way to generate a unique value each time.)

jhadvig replied:
@spadgett then I think we need to define the validation rules as:
- Hash directive format: Check for a valid hash algorithm (like SHA-256) followed by a hyphen and a valid Base64 value.
- Host source format: Check that the input matches the pattern of a valid URL.

For the CEL rule below accomplishes this by checking for two conditions:

- The value starts with a valid hash algorithm (sha256, sha384, or sha512) followed by a - and a valid Base64 string.
- The value matches the pattern of a valid host source (beginning with http:// or https://).

WDYT ?
ccing @vojtechszocs

spadgett replied:
Host sources don't need to start with `http://` or `https://`. For example, `*.example.com` is a valid host source.

jhadvig replied:
Right, sorry. I've unintentionally leaved it out, but it should be valid as well 👍

vojtechszocs replied:
@spadgett @jhadvig I think the currently proposed validation rules
```
// +kubebuilder:validation:MinLength=1
// +kubebuilder:validation:MaxLength=1024
// +kubebuilder:validation:XValidation:rule="!self.contains(\"'\")",message="CSP directive value cannot contain a quote"
// +kubebuilder:validation:XValidation:rule="!self.matches('\\\\s')",message="CSP directive value cannot contain a whitespace"
// +kubebuilder:validation:XValidation:rule="!self.contains(',')",message="CSP directive value cannot contain a comma"
// +kubebuilder:validation:XValidation:rule="!self.contains(';')",message="CSP directive value cannot contain a semi-colon"
// +kubebuilder:validation:XValidation:rule="self != '*'",message="CSP directive value cannot be a wildcard"
type CSPDirectiveValue string
```
are reasonably strict, we can always loosen these rules based on CSP violation reports once we start collecting them with telemetry.

Additional validation logic can be placed into the Console operator (since it's responsible for ensuring valid configuration of Console and all of its resources) or directly into Bridge [`server.go`](https://github.com/openshift/console/blob/master/pkg/server/server.go) function `indexHandler` where we generate the actual CSP string.

One example of additional validation would be e.g. `DefaultSrc` and `ScriptSrc` containing the same value, in which case we want to remove that value from `ScriptSrc` since the `DefaultSrc` directive acts as a fallback for the `ScriptSrc` directive.

Another example of additional validation would be to ensure the plugin provided CSP source values are valid URLs with potential `*` wildcards.

The only `'something'` exception that makes sense to me is [`'<hash-algorithm>-<base64-value>'`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources#hash-algorithm-base64-value) but I think it's better to wait for CSP violation reports first before we start allowing such exceptions. (Other `'something'` special values don't seem too relevant to me at the moment, the main focus should be allowing additional plugin-specific remote sources so they can be loaded without issues in the browser.)

jhadvig replied:
I would agree that lets start with stricter validation and relax it down the road, since we will be in report mode for couple of releases.

JoelSpeed replied:
> That said I think we should disable the quote completely. If you agree will also keep the previous check for a value which starts and ends with quote?

If we are removing quotes completely, why would we need a separate check to check for starting and ending with quotes? 

If we need to allow the start and end quotes specifically just for the hash algorithm, we could relax the no quotes validation in a compatible way later I believe.

jhadvig replied:
@JoelSpeed sorry for the confusion, I've removed the the check for starting and ending with quotes as well in the end 👍

spadgett replied:
I am good with disallowing single quotes to start and relaxing validation later if needed

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,130 @@ type ConsolePluginSpec struct {
+
+// DirectiveType is an enumeration of OpenShift web console supported CSP directives.
+// LoadType is an enumeration of i18n loading types.
+// +kubebuilder:validation:Enum:="DefaultSrc";"ScriptSrc";"StyleSrc";"ImgSrc";"FontSrc"
+type DirectiveType string
```
Can you also add the following, it adds additional validations to the openapi schema we generate
```suggestion
// +kubebuilder:validation:Enum:="DefaultSrc";"ScriptSrc";"StyleSrc";"ImgSrc";"FontSrc"
// +enum
type DirectiveType string
```

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,130 @@ type ConsolePluginSpec struct {
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x == y))",message="each CSP directive value must be unique"
+	// +listType=set
```
Set doesn't work with SSA, use atomic
```suggestion
	// +listType=atomic
```

JoelSpeed reviewed (commented):
I believe the CRD generation needs to be re-run please

jhadvig commented:
/retest

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,131 @@ type ConsolePluginSpec struct {
+	// values defines an array of values to append to the console defaults for this directive.
+	// Each ConsolePlugin may define their own directives with their values. These will be set
+	// by the OpenShift web console's backend, as part of its Content-Security-Policy header.
+	// The array can contain at most 16 values. Each directive value must have a maximum length
+	// of 1024 characters and must not contain whitespace, commas, semicolons. The value '*' or
+	// strings enclosed in single quotes are not permitted.
```
We've changed this validation slightly since this was written
```suggestion
	// The array can contain at most 16 values. Each directive value must have a maximum length
	// of 1024 characters and must not contain whitespace, commas (,), semicolons (;) or single quotes (').
	// The value '*' is not permitted.
```

spadgett replied:
@JoelSpeed API looks good to me once we address this last comment. Thanks, Joel.

JoelSpeed reviewed (commented):
Other than this, I'm good with the API now

@spadgett Are you happy with the validations we are imposing here? Any feedback on the API?

jhadvig commented:
/retest

spadgett reviewed (commented):
LGTM

Deferring to @JoelSpeed for final approval

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -48,6 +48,131 @@ type ConsolePluginSpec struct {
+	//   OpenShift web console server CSP response header:
+	//     Content-Security-Policy: default-src 'self'; base-uri 'self'; script-src 'self' https://script1.com/ https://script2.com/ https://script3.com/; font-src 'self' https://font1.com/ https://font2.com/; img-src 'self' https://img1.com/; style-src 'self'; frame-src 'none'; object-src 'none'
+	//
+	// +openshift:enable:FeatureGate=ConsolePluginCSP
```
```suggestion
	// +openshift:enable:FeatureGate=ConsolePluginContentSecurityPolicy
```

JoelSpeed commented:
/lgtm

jhadvig commented:
/retest

jhadvig commented:
/test e2e-aws-serial

JoelSpeed commented on console/v1/types_console_plugin.go:
```diff
@@ -43,11 +43,138 @@ type ConsolePluginSpec struct {
 	Backend ConsolePluginBackend `json:"backend"`
 	// proxy is a list of proxies that describe various service type
 	// to which the plugin needs to connect to.
+	// +listType=map
+	// +listMapKey=alias
```
Since this is in spec, and you don't otherwise have anything validating that the proxy aliases in the list are unique, I think this list will need to be marked atomic

If we were already validating uniqueness of the aliases, then this would be correct, but I don't see anything that prevents anyone having duplicate aliases right now

JoelSpeed commented:
/lgtm
/override ci/prow/verify-crd-schema

Only errors refer to unfixable errors or those in the v1alpha1 schema which is meant to have been dropped by now

jhadvig commented:
/retest

JoelSpeed commented:
/test e2e-aws-serial

JoelSpeed commented:
/hold cancel

JoelSpeed commented:
/override ci/prow/verify-crd-schema
